libyul/optimiser/BlockHasher.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimiser component that calculates hash values for block prefixes.
+ */
+
+#include <libyul/optimiser/BlockHasher.h>
+#include <libyul/optimiser/SyntacticalEquality.h>
+#include <libyul/Utilities.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+namespace {
+static constexpr uint64_t compileTimeLiteralHash(char const* _literal, size_t _N)
+{
+	return (_N == 0) ? 14695981039346656037u : compileTimeLiteralHash(_literal + 1, _N - 1) ^ 1099511628211u;
```

Comment:
Where are these numbers taken from?

Reply:
Those are the FNV primes - they should maybe be defined globally somewhere, they are used in ``BlockHasher.h`` and in ``YulString.h`` as well...
More interestingly, I'm just realizing, that the recursion is completely wrong, I don't know how that happened :-). I'll fix it.